### PR TITLE
icon layer

### PR DESCRIPTION
**Pull request by simon-contreras-deel**
https://app.clubhouse.io/cartoteam/story/70649/styles-icon-layer

**API**
With icon atlas:
```
const ICON_MAPPING = {
  marker: { x: 0, y: 0, width: 128, height: 128, mask: true }
};

const style = carto.viz.iconStyle({
  iconAtlas: 'https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/icon-atlas.png',
  iconMapping: ICON_MAPPING,
  getIcon: d => 'marker',
});
```

with custom URL icon:
```
const style = carto.viz.iconStyle({
  getIcon: () => ({
    url: 'https://avatars2.githubusercontent.com/u/421145?s=60&v=4',
    height: 60,
    width: 60
  })
});
```

**Usage**:
```
const style = carto.viz.iconStyle({
  getIcon: () => ({
    url: IMAGE_URL,
    height: 60,
    width: 60
  })
});

const layer = new carto.viz.Layer('table', style);
layer.addTo(deckMap);
```

Example in `examples/_debug/styles/icon-style.html`

### CODE REVIEW

**Comment by simon-contreras-deel** [comment]
Finally the API is:
```
const style = carto.viz.style.icon('./carto.jpg')
```

You can use a options object:
```
const style = carto.viz.style.icon('./carto.jpg', {
  iconHeight: 30,
  iconWidth: 30
});
```

Other options are `anchorX`, `anchorY` and `sizeScale`

File: src/lib/viz/style/helpers/icon-style.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
+import { IconLayer } from '@deck.gl/layers';
+import { Feature, Point, MultiPoint } from 'geojson';
+import { GeometryType } from '@/viz/source';
+import { StyledLayer } from '../layer-style';
+import { Style, getStyles } from '..';
+import { CartoStylingError, stylingErrorTypes } from '../../errors/styling-error';
+
+export interface IconOptionsStyle {
+  // icon width in pixels
+  iconWidth: number;
```
Comment (VictorVelarde) [inline src/lib/viz/style/helpers/icon-style.ts]: small suggestion on renaming this to just 'width' and 'height'

File: src/lib/viz/style/helpers/icon-style.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  iconHeight: number;
+  // horizontal position of icon anchor
+  anchorX: number;
+  // vertical position of icon anchor
+  anchorY: number;
+  // Icon size multiplier
+  sizeScale: number;
+}
+
+interface IconOptionsStyleValidated {
```
Comment (VictorVelarde) [inline src/lib/viz/style/helpers/icon-style.ts:21]: Why this 'Validated' suffix? Have we done something like this in other style helpers?

File: examples/style-helpers/icon-style.html [hunk 3]
```diff
@@ -0,0 +1,73 @@
+<!DOCTYPE html>
```
Comment (VictorVelarde) [inline examples/style-helpers/icon-style.html:1]: Please adapt this example to the template, to have the airship Panel for the title & description

File: src/lib/viz/style/helpers/icon-style.ts (hunk 2 above)
Comment (simon-contreras-deel) [reply src/lib/viz/style/helpers/icon-style.ts:21]: TS things,it is the easiest solution to avoid errors

File: src/lib/viz/layer/Layer.ts [hunk 4]
```diff
@@ -266,7 +266,11 @@ export class Layer extends WithEvents implements StyledLayer {
     if (this._source.sourceType === 'SQLSource' || this._source.sourceType === 'DatasetSource') {
       this._deckLayer = new MVTLayer(layerProperties);
     } else if (this._source.sourceType === 'GeoJSONSource') {
-      this._deckLayer = new GeoJsonLayer(layerProperties);
+      if (layerProperties._isIconLayer) {
+        this._deckLayer = new IconLayer(layerProperties);
```
Comment (VictorVelarde) [inline src/lib/viz/layer/Layer.ts:270]: I guess this doesn't have any effect on things related to getting the features and so on, right? For example, have you tried using local dataviews on Layers with IconLayer?

**Review by VictorVelarde — APPROVED** [review]
Nice work! Just small comments and questions, feel free to merge

**Comment by VictorVelarde** [comment]
Ups, I forgot a thing: add a line to the CHANGELOG!

File: src/lib/viz/layer/Layer.ts (hunk 4 above)
Comment (simon-contreras-deel) [reply src/lib/viz/layer/Layer.ts:270]: Good point! No, I have not

Comment (simon-contreras-deel) [reply src/lib/viz/layer/Layer.ts:270]: it is failing with geojson + icon + dataview filtered

Comment (VictorVelarde) [reply src/lib/viz/layer/Layer.ts:270]: then let's fix that! 😄 

After dataviews, interactivity will be the next piece to test!

Comment (simon-contreras-deel) [reply src/lib/viz/layer/Layer.ts:270]: done and checked

**Pull request merged**